### PR TITLE
ADCM-1409 UI: new module for organizing test data

### PR DESCRIPTION
**Pull request by DavidovPavel**

### CODE REVIEW

File: web/src/app/core/types/func.ts [hunk 1]
```diff
@@ -83,20 +60,34 @@ export function flatten<T>(a: T[]) {
...
  * @returns 16-bit hex string
  * @example '#110C2E'
  */
 export function getRandomColor(): string {
   const letters = '0123456789ABCDEF';
-  return newArray(6).reduce((p) => (p += letters[Math.floor(Math.random() * 16)]), '#');
+  return newArray(6, (_, i) => i).reduce((p) => (p += letters[Math.floor(Math.random() * 16)]), '#');
 }
 
 export function uniqid(prefix: string = '', isFloat: boolean = false) {
```
Comment (drKaramazin) [inline web/src/app/core/types/func.ts:84]: need to declare a return type

File: web/src/app/core/types/host-component.ts [hunk 2]
```diff
@@ -38,7 +38,7 @@ export interface HostComponent {
 /**
  * A service component that may be installed on hosts in a cluster.
  */
-export interface Component {
+export interface IComponent {
```
Comment (drKaramazin) [inline web/src/app/core/types/host-component.ts:41]: What criterion do you choose for the prefix of an interface?

File: web/src/app/shared/configuration/field.service.ts [hunk 3]
```diff
@@ -36,81 +36,115 @@ export interface IToolsEvent {
...
+const patternFn = {
+  integer: intPattern,
+  int: intPattern,
+  float: () => new RegExp(/^[-]?[0-9]+(\.[0-9]+)?$/),
+};
+
+export const getPattern = (t: TNForm): RegExp => (patternFn[t] ? patternFn[t]() : null);
+
+const fn = {
+  boolean: (v: boolean | null, d: boolean | null, r: boolean) => (String(v) === 'true' || String(v) === 'false' || String(v) === 'null' ? v : r ? d : null),
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/field.service.ts:65]: need to declare a return type

File: web/src/app/shared/configuration/field.service.ts [hunk 4]
```diff
@@ -36,81 +36,115 @@ export interface IToolsEvent {
...
+  json: (v: string) => (v === null ? '' : JSON.stringify(v, undefined, 4)),
+  map: (v: object, d: object) => (!v ? d : v),
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/field.service.ts:67]: need to declare a return type

File: web/src/app/shared/configuration/field.service.ts [hunk 5]
```diff
@@ -36,81 +36,115 @@ export interface IToolsEvent {
...
+  list: (v: string[], d: string[]) => (!v ? d : v),
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/field.service.ts:68]: need to declare a return type

File: web/src/app/shared/configuration/field.service.ts [hunk 6]
```diff
@@ -36,81 +36,115 @@ export interface IToolsEvent {
...
+  structure: (v: any) => v,
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/field.service.ts:69]: need to declare a return type

File: web/src/app/shared/configuration/field.service.ts [hunk 7]
```diff
@@ -36,81 +36,115 @@ export interface IToolsEvent {
...
+};
+
+export const getValue = (t: TNForm) => {
+  const def = (value: number | string) => (value === null || value === undefined ? '' : String(value));
+  return fn[t] ? fn[t] : def;
+};
+
+export const getKey = (n: string, sn: string) => (sn ? `${sn}/${n}` : n);
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/field.service.ts:77]: I don't understand the mean of this function.

File: web/src/app/shared/configuration/field.service.ts [hunk 8]
```diff
@@ -36,81 +36,115 @@ export interface IToolsEvent {
...
+  ...item,
+  key: getKey(item.name, item.subname),
+  value: getValue(item.type)(item.value, item.default, item.required),
+  validator: getValidator(item.required, item.limits?.min, item.limits?.max, item.type),
+  controlType: getControlType(item.type),
+  hidden: item.name === '__main_info' || isHidden(item),
+  compare: [],
+});
+
+const fo = (n: string) => (b: IFieldStack) => b.type !== 'group' && b.subname && b.name === n;
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/field.service.ts:96]: I don't see the meaning of this function. It uses only one time

File: web/src/app/shared/configuration/field.service.ts [hunk 9]
```diff
@@ -36,81 +36,115 @@ export interface IToolsEvent {
...
+    .map((c) => ({ ...c, name: c.subname, activatable: a.activatable && !isActive(d.attr, a.name) }));
+
+const getPanel = (a: IFieldStack, d: IConfig): IPanelOptions => ({
+  ...a,
+  hidden: isHidden(a),
+  active: a.activatable ? isActive(d.attr, a.name) : true,
+  options: getOptions(a, d),
+});
+
+const handleTree = (c: ISearchParam) => (a: TFormOptions) => {
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/field.service.ts:112]: need to declare a return type

File: web/src/app/shared/configuration/field.service.ts [hunk 10]
```diff
@@ -183,34 +217,21 @@ export class FieldService {
...
 
   /**
    * Output form, cast to source type
    */
   public parseValue(output: IOutput, source: ISource[]): IOutput {
-    const findField = (name: string, p?: string): Partial<FieldStack> => source.find((a) => (p ? a.name === p && a.subname === name : a.name === name) && !a.read_only);
+    const findField = (name: string, p?: string): Partial<IFieldStack> => source.find((a) => (p ? a.name === p && a.subname === name : a.name === name) && !a.read_only);
 
-    const runYspecParse = (v: any, f: Partial<FieldOptions>) => ((v === {} || v === []) && !f.value ? f.value : this.runYspec(v, f.limits.rules));
+    const runYspecParse = (v: any, f: Partial<IFieldOptions>) => ((v === {} || v === []) && !f.value ? f.value : this.runYspec(v, f.limits.rules));
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/field.service.ts:230]: It seems to me this row have a bug

File: web/src/app/shared/configuration/fields/fields.component.ts [hunk 11]
```diff
@@ -57,14 +57,14 @@ export class ConfigFieldsComponent {
...
+    return this.dataOptions.filter((a) => a.type === 'group' && (a as IPanelOptions).activatable).reduce((p, c: IPanelOptions) => ({ ...p, [c.name]: { active: c.active } }), {});
   }
 
-  isPanel(item: FieldOptions | PanelOptions) {
+  isPanel(item: TFormOptions) {
     return 'options' in item && !item.hidden;
   }
 
-  trackBy(index: number, item: PanelOptions): string {
+  trackBy(index: number, item: IPanelOptions): string {
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/fields/fields.component.ts:67]: It seems that the index is not needed

File: web/src/app/shared/configuration/scheme/scheme.service.spec.ts [hunk 12]
```diff
@@ -14,10 +14,10 @@ import { TestBed } from '@angular/core/testing';
 import { SchemeService } from './scheme.service';
 import { FieldService } from '../field.service';
 import { FormGroup, FormControl, FormArray } from '@angular/forms';
-import { FieldOptions } from '../types';
+import { IFieldOptions } from '../types';
 
 const form = new FormGroup({ field: new FormControl() });
-const field = (<unknown>{ display_name: 'field_display_name', name: 'field', limits: {}, required: true, value: null, default: null }) as FieldOptions;
+const field = (<unknown>{ display_name: 'field_display_name', name: 'field', limits: {}, required: true, value: null, default: null }) as IFieldOptions;
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/scheme/scheme.service.spec.ts:20]: Why is the field's type unknown?

File: web/src/app/shared/configuration/tests/configuration.ts [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import { getControlType, getKey, getOptions, getValidator, getValue, TFormOptions, IOutput } from '../field.service';
+
+import { IConfig, IConfigAttr, IFieldStack, ILimits, IUIoptions, stateType, TNForm, TValue } from '../types';
+import { IYContainer, IYField, IYspec } from '../yspec/yspec.service';
+
+export class YContainer {}
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/tests/configuration.ts:17]: It seems useless

File: web/src/app/shared/configuration/tests/configuration.ts [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+export class YField {}
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/tests/configuration.ts:19]: It seems useless

File: web/src/app/shared/configuration/tests/configuration.ts [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+
+export class Yspec {
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/tests/configuration.ts:21]: It seems useless

File: web/src/app/shared/configuration/tests/configuration.ts [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+  constructor() {}
+}
+
+export class UIOptions implements IUIoptions {}
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/tests/configuration.ts:25]: It seems useless

File: web/src/app/shared/configuration/tests/configuration.ts [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+
+export class Limits {
+  min: number;
+  max: number;
+  option: any;
+  read_only: stateType[];
+  active: boolean;
+  rules: IYField | IYContainer;
+  constructor(public yspec: IYspec) {}
+  set Rules(rules: IYField | IYContainer) {
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/tests/configuration.ts:35]: It seems useless. Why does the name start with a capital letter?

File: web/src/app/shared/configuration/tests/configuration.ts [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+const toPanel = (a: IFieldStack, data: IConfig) => ({
+  ...a,
+  options: getOptions(a, data),
+  active: true,
+  hidden: false,
+});
+
+const toField = (a: IFieldStack) => ({
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/tests/configuration.ts:93]: need to declare a return type

File: web/src/app/shared/configuration/tests/configuration.ts [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+
+  /**
+   * return group if params as array
+   */
+  public static add(params: (TNForm | TNForm[])[]) {
+    return params.reduce<IFieldStack[]>((p, c, i) => [...p, ...(Array.isArray(c) ? this.addGroup(i, c) : [new FieldStack(i, c)])], []);
+  }
+}
+
+const toPanel = (a: IFieldStack, data: IConfig) => ({
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/tests/configuration.ts:86]: need to declare a return type

File: web/src/app/shared/configuration/tests/configuration.ts [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+  }, []);
+};
+
+export const setValue = (data: IFieldStack[], v: TValue): IOutput =>
+  data
+    .filter((a) => !a.subname)
+    .reduce(
+      (p, c, i) => ({
+        ...p,
+        [c.name]: c.type === 'group' ? data.filter((a) => a.name === c.name && a.type !== 'group').reduce((a, b, k) => ({ ...a, [b.subname]: v[i][k] }), {}) : v ? v[i] : null,
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/tests/configuration.ts:117]: This line is too long

File: web/src/app/shared/configuration/tests/configuration.ts [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+
+export class FieldStack implements IFieldStack {
+  subname: string;
+  display_name: string;
+  default: TValue = null;
+  description?: string;
+  limits: ILimits;
+  ui_options: IUIoptions;
+
+  constructor(id: number, public type: TNForm, public name: string = null, public value = null, public required = true, public read_only = false, public activatable = false) {
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/tests/configuration.ts:48]: This line is too long

File: web/src/app/shared/configuration/tests/configuration.ts (hunk 20 above)
Comment (drKaramazin) [reply web/src/app/shared/configuration/tests/configuration.ts:117]: The logic is not obvious

File: web/src/app/shared/host-components-map/services2hosts/service-host.component.spec.ts [hunk 22]
```diff
@@ -13,189 +13,34 @@ import { NO_ERRORS_SCHEMA } from '@angular/core';
...
-  ],
-};
+function genData() {
+  const _c = ComponentFactory(4, 1);
+  const host = [new HcmHost('test', 1)];
+  return { component: _c, host, hc: HCFactory(1, 1, 4) };
+}
+
+function domFirstElems(n: HTMLElement) {
+  const components = n.querySelectorAll('.wrapper').item(0).querySelectorAll('app-much-2-many');
```
Comment (drKaramazin) [inline web/src/app/shared/host-components-map/services2hosts/service-host.component.spec.ts:36]: needs to declare a type

File: web/src/app/shared/host-components-map/services2hosts/service-host.component.spec.ts [hunk 23]
```diff
@@ -13,189 +13,34 @@ import { NO_ERRORS_SCHEMA } from '@angular/core';
...
+  const hosts = n.querySelectorAll('.wrapper').item(1).querySelectorAll('app-much-2-many');
```
Comment (drKaramazin) [inline web/src/app/shared/host-components-map/services2hosts/service-host.component.spec.ts:37]: needs to declare a type

File: web/src/app/shared/host-components-map/services2hosts/service-host.component.spec.ts [hunk 24]
```diff
@@ -13,189 +13,34 @@ import { NO_ERRORS_SCHEMA } from '@angular/core';
...
+  const host = hosts.item(0);
```
Comment (drKaramazin) [inline web/src/app/shared/host-components-map/services2hosts/service-host.component.spec.ts:38]: needs to declare a type

File: web/src/app/shared/host-components-map/services2hosts/service-host.component.spec.ts [hunk 25]
```diff
@@ -13,189 +13,34 @@ import { NO_ERRORS_SCHEMA } from '@angular/core';
...
+  const comp = components.item(0);
```
Comment (drKaramazin) [inline web/src/app/shared/host-components-map/services2hosts/service-host.component.spec.ts:39]: needs to declare a type

File: web/src/app/shared/host-components-map/services2hosts/service-host.component.ts [hunk 26]
```diff
@@ -235,11 +236,15 @@ export class ServiceHostComponent extends SocketListenerDirective implements OnI
       a.isLink = false;
       a.relations = [];
     };
-    this.statePost.clear();
-    this.statePost.update(this.loadPost.data);
+    
     this.Hosts.forEach(ma);
```
Comment (drKaramazin) [inline web/src/app/shared/host-components-map/services2hosts/service-host.component.ts:240]: What is the "ma"?

File: web/src/app/shared/host-components-map/services2hosts/service-host.component.ts [hunk 27]
```diff
@@ -235,11 +236,15 @@ export class ServiceHostComponent extends SocketListenerDirective implements OnI
...
     this.Components.forEach(ma);
```
Comment (drKaramazin) [inline web/src/app/shared/host-components-map/services2hosts/service-host.component.ts:241]: What is the "ma"?

File: web/src/app/shared/host-components-map/take.service.ts [hunk 28]
```diff
@@ -13,25 +13,60 @@ import { Injectable } from '@angular/core';
...
 
 import { AddService } from '../add-component/add.service';
 import { DialogComponent } from '../components';
 import { DependenciesComponent } from './dependencies.component';
-import { CompTile, Constraint, HostTile, IRawHosComponent, IStream, Post, StatePost, Tile } from './types';
+import { CompTile, TConstraint, HostTile, IRawHosComponent, IStream, Post, StatePost, Tile } from './types';
 
 export const getSelected = (from: Tile[]) => from.find((a) => a.isSelected);
 
+const isShrink = (ap: IActionParameter[]) => ap.every((a) => a.action === 'remove');
```
Comment (drKaramazin) [inline web/src/app/shared/host-components-map/take.service.ts:26]: needs to declare a return type

File: web/src/app/shared/host-components-map/take.service.ts [hunk 29]
```diff
@@ -13,25 +13,60 @@ import { Injectable } from '@angular/core';
...
+const isExpand = (ap: IActionParameter[]) => ap.every((a) => a.action === 'add');
```
Comment (drKaramazin) [inline web/src/app/shared/host-components-map/take.service.ts:27]: needs to declare a return type

File: web/src/app/shared/host-components-map/take.service.ts [hunk 30]
```diff
@@ -13,25 +13,60 @@ import { Injectable } from '@angular/core';
...
+const accord = (a: IActionParameter) => (b: CompTile) => b.component === `${a.service}/${a.component}`;
```
Comment (drKaramazin) [inline web/src/app/shared/host-components-map/take.service.ts:28]: needs to declare a return type

File: web/src/app/shared/host-components-map/take.service.ts [hunk 31]
```diff
@@ -13,25 +13,60 @@ import { Injectable } from '@angular/core';
...
+const existCondition = (rel: CompTile[], ap: IActionParameter[]) =>
```
Comment (drKaramazin) [inline web/src/app/shared/host-components-map/take.service.ts:29]: needs to declare a return type

File: web/src/app/shared/host-components-map/take.service.ts [hunk 32]
```diff
@@ -13,25 +13,60 @@ import { Injectable } from '@angular/core';
...
+  isShrink(ap) ? ap.some((a) => rel.some(accord(a))) : ap.every((a) => rel.some(accord(a)));
+const checkEmptyHost = (h: HostTile, ap: IActionParameter[]) =>
```
Comment (drKaramazin) [inline web/src/app/shared/host-components-map/take.service.ts:31]: needs to declare a return type

File: web/src/app/shared/host-components-map/take.service.ts [hunk 33]
```diff
@@ -13,25 +13,60 @@ import { Injectable } from '@angular/core';
...
+  ap ? (existCondition(h.relations as CompTile[], ap) ? isExpand(ap) : isShrink(ap)) : false;
+
+//#region user click
+
+const checkConstraint = (c: TConstraint, r: number) => {
```
Comment (drKaramazin) [inline web/src/app/shared/host-components-map/take.service.ts:36]: needs to declare a return type

File: web/src/app/shared/host-components-map/take.service.ts [hunk 34]
```diff
@@ -13,25 +13,60 @@ import { Injectable } from '@angular/core';
...
+  if (!c?.length) return true;
+  const v = c[c.length - 1];
+  return v === '+' || v === 'odd' || v > r || v === 'depend';
+};
+
+const flag = (host_id: number, com: CompTile, load: StatePost) =>
```
Comment (drKaramazin) [inline web/src/app/shared/host-components-map/take.service.ts:42]: needs to declare a return type

File: web/src/app/shared/host-components-map/take.service.ts [hunk 35]
```diff
@@ -13,25 +13,60 @@ import { Injectable } from '@angular/core';
...
+  load.data.some((a) => a.component_id === com.id && a.service_id === com.service_id && a.host_id === host_id);
+
+const checkActions = (host_id: number, com: CompTile, action: 'add' | 'remove', load: StatePost): boolean => {
+  if (com.actions?.length) {
+    if (action === 'remove') return flag(host_id, com, load) ? com.actions.some((a) => a === 'remove') : true;
+    if (action === 'add') return flag(host_id, com, load) ? true : com.actions.some((a) => a === 'add');
+  } else return true;
+};
+
+const findDependencies = (c: CompTile, cs: CompTile[]) => {
```
Comment (drKaramazin) [inline web/src/app/shared/host-components-map/take.service.ts:52]: needs to declare a return type

File: web/src/app/shared/host-components-map/take.service.ts [hunk 36]
```diff
@@ -13,25 +13,60 @@ import { Injectable } from '@angular/core';
...
+  const r =
+    c.requires?.reduce((p, a) => [...p, ...a.components.map((b) => ({ prototype_id: b.prototype_id }))], []) || [];
+  return cs.filter((a) => r.some((b) => b.prototype_id === a.prototype_id));
+};
+
+const checkDependencies = (c: CompTile, cs: CompTile[]) =>
```
Comment (drKaramazin) [inline web/src/app/shared/host-components-map/take.service.ts:58]: needs to declare a return type

File: web/src/app/shared/host-components-map/take.service.ts [hunk 37]
```diff
@@ -13,25 +13,60 @@ import { Injectable } from '@angular/core';
...
+  findDependencies(c, cs).forEach((a) => (a.limit = a.limit ? [...a.limit, 'depend'] : ['depend']));
+
+const checkRequires = (component: CompTile, cs: CompTile[]) =>
```
Comment (drKaramazin) [inline web/src/app/shared/host-components-map/take.service.ts:61]: needs to declare a return type

File: web/src/app/shared/host-components-map/test/index.ts [hunk 38]
```diff
@@ -0,0 +1,58 @@
...
+  requires: HCmRequires[];
+  constructor(public id: number, public service_id: number) {
+    this.prototype_id = id;
+    this.name = `component_${id}`;
+    this.display_name = `component_display_name_${id}`;
+    this.service_name = `service_${service_id}`;
+  }
+}
+
+export const ComponentFactory = (count: number, service_id: number) =>
```
Comment (drKaramazin) [inline web/src/app/shared/host-components-map/test/index.ts:54]: needs to declare a return type

File: web/src/app/shared/host-components-map/test/index.ts [hunk 39]
```diff
@@ -0,0 +1,58 @@
...
+  newArray<IComponent>(count, (_, i) => new HCmComponent(i + 1, service_id) as IComponent);
+
+export const HCFactory = (host_id: number, service_id: number, components: number): Post[] =>
```
Comment (drKaramazin) [inline web/src/app/shared/host-components-map/test/index.ts:57]: needs to declare a return type

**Review by drKaramazin — CHANGES_REQUESTED** [review]

File: web/src/app/shared/configuration/field.service.ts (hunk 10 above)
Comment (drKaramazin) [reply web/src/app/shared/configuration/field.service.ts:230]: 1). any "v" !== {} and !== []
2). "(v === {} || v === [])" === false
3). "(v === {} || v === []) && !f.value" always === false

**Review by drKaramazin — CHANGES_REQUESTED** [review]

File: web/src/app/core/types/host-component.ts (hunk 2 above)
Comment (DavidovPavel) [reply web/src/app/core/types/host-component.ts:41]: I - interface
UpperCase for class

File: web/src/app/shared/configuration/fields/fields.component.ts (hunk 11 above)
Comment (DavidovPavel) [reply web/src/app/shared/configuration/fields/fields.component.ts:67]: https://angular.io/guide/structural-directives#inside-ngfor

File: web/src/app/shared/configuration/scheme/scheme.service.spec.ts (hunk 12 above)
Comment (DavidovPavel) [reply web/src/app/shared/configuration/scheme/scheme.service.spec.ts:20]: This is mock for test

File: web/src/app/shared/configuration/tests/configuration.ts (hunk 13 above)
Comment (DavidovPavel) [reply web/src/app/shared/configuration/tests/configuration.ts:17]: it' todo list

File: web/src/app/core/store/sockets/socket.reducer.ts [hunk 40]
```diff
@@ -12,6 +12,20 @@
...
+    type: TypeName;  // string 'cluster' | 'host' ...etc.
+    id: number;
+    details: {
+      id?: string;
+      type: string;
+      value: any;
+    }
+   }
+   ```
+ */
```
[the remark this answers is not in the document]
Comment (DavidovPavel) [reply web/src/app/core/store/sockets/socket.reducer.ts:17]: Comments as possible

File: web/src/app/core/types/api.ts [hunk 41]
```diff
@@ -10,26 +10,20 @@
...
-  | 'service'
-  | 'job'
-  | 'task'
-  | 'user'
-  | 'profile'
-  | 'adcm'
-  | 'stats'
-  | 'hostcomponent'
-  | 'component';
+export type TypeName = 'bundle' | 'cluster' | 'host' | 'provider' | 'service' | 'job' | 'task' | 'user' | 'profile' | 'adcm' | 'stats' | 'hostcomponent' | 'component';
```
[the remark this answers is not in the document]
Comment (DavidovPavel) [reply web/src/app/core/types/api.ts:17]: not necessary in this case

File: web/src/app/shared/configuration/field.service.ts (hunk 8 above)
Comment (DavidovPavel) [reply web/src/app/shared/configuration/field.service.ts:96]: just function as argument for filter by name for config field

File: web/src/app/shared/configuration/field.service.ts (hunk 9 above)
Comment (DavidovPavel) [reply web/src/app/shared/configuration/field.service.ts:112]: done

File: web/src/app/shared/configuration/field.service.ts (hunk 10 above)
Comment (DavidovPavel) [reply web/src/app/shared/configuration/field.service.ts:230]: fixed

File: web/src/app/shared/configuration/tests/configuration.ts (hunk 13 above)
Comment (DavidovPavel) [reply web/src/app/shared/configuration/tests/configuration.ts:17]: it' todo list

File: web/src/app/shared/configuration/tests/configuration.ts (hunk 14 above)
Comment (DavidovPavel) [reply web/src/app/shared/configuration/tests/configuration.ts:19]: it' todo list

File: web/src/app/shared/configuration/tests/configuration.ts (hunk 16 above)
Comment (DavidovPavel) [reply web/src/app/shared/configuration/tests/configuration.ts:25]: it' todo list

File: web/src/app/shared/configuration/tests/configuration.ts (hunk 17 above)
Comment (DavidovPavel) [reply web/src/app/shared/configuration/tests/configuration.ts:35]: it' todo list

File: web/src/app/shared/configuration/tests/configuration.ts (hunk 21 above)
Comment (DavidovPavel) [reply web/src/app/shared/configuration/tests/configuration.ts:48]: done

File: web/src/app/shared/configuration/tests/configuration.ts (hunk 19 above)
Comment (DavidovPavel) [reply web/src/app/shared/configuration/tests/configuration.ts:86]: done

File: web/src/app/shared/configuration/tests/configuration.ts (hunk 18 above)
Comment (DavidovPavel) [reply web/src/app/shared/configuration/tests/configuration.ts:93]: done

File: web/src/app/shared/host-components-map/test/index.ts (hunk 39 above)
Comment (DavidovPavel) [reply web/src/app/shared/host-components-map/test/index.ts:57]: Post[]

File: web/src/app/shared/host-components-map/test/index.ts (hunk 38 above)
Comment (DavidovPavel) [reply web/src/app/shared/host-components-map/test/index.ts:54]: done

File: web/src/app/shared/configuration/field.service.ts (hunk 8 above)
Comment (drKaramazin) [reply web/src/app/shared/configuration/field.service.ts:109]: Just will name the function like "filterByName"

File: web/src/app/shared/configuration/field.service.ts [hunk 42]
```diff
@@ -36,81 +48,126 @@ export interface IToolsEvent {
...
+    .map((c) => ({ ...c, name: c.subname, activatable: a.activatable && !isActive(d.attr, a.name) }));
+
+const getPanel = (a: IFieldStack, d: IConfig): IPanelOptions => ({
+  ...a,
+  hidden: isHidden(a),
+  active: a.activatable ? isActive(d.attr, a.name) : true,
+  options: getOptions(a, d),
+});
+
+const handleTree = (c: ISearchParam): ((a: TFormOptions) => TFormOptions) => (a: TFormOptions): TFormOptions => {
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/field.service.ts:125]: I couldn't agree with this type

File: web/src/app/shared/configuration/field.service.ts [hunk 43]
```diff
@@ -36,81 +48,126 @@ export interface IToolsEvent {
...
+  if ('options' in a) {
+    a.options = a.options.map(handleTree(c));
```
Comment (drKaramazin) [inline web/src/app/shared/configuration/field.service.ts:127]: It can invoke many problems

**Review by drKaramazin — CHANGES_REQUESTED** [review]

File: web/src/app/shared/configuration/tests/configuration.ts (hunk 13 above)
Comment (drKaramazin) [reply web/src/app/shared/configuration/tests/configuration.ts:18]: Please, comment it like 'TODO' in the code

File: web/src/app/shared/configuration/tests/configuration.ts (hunk 14 above)
Comment (drKaramazin) [reply web/src/app/shared/configuration/tests/configuration.ts:20]: Please, comment it like 'TODO' in the code

File: web/src/app/shared/configuration/tests/configuration.ts (hunk 15 above)
Comment (drKaramazin) [reply web/src/app/shared/configuration/tests/configuration.ts:22]: Please, comment it like 'TODO' in the code

File: web/src/app/shared/configuration/tests/configuration.ts (hunk 16 above)
Comment (drKaramazin) [reply web/src/app/shared/configuration/tests/configuration.ts:26]: Please, comment it like 'TODO' in the code

File: web/src/app/shared/configuration/tests/configuration.ts (hunk 17 above)
Comment (drKaramazin) [reply web/src/app/shared/configuration/tests/configuration.ts:36]: Please, comment it like 'TODO' in the code

**Review by acmnu — APPROVED** [review]

**Pull request merged**